lib/NodeInstantiator.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\RST;
+
+use Doctrine\RST\Nodes\Node;
+use InvalidArgumentException;
+use ReflectionClass;
+use function sprintf;
+
+class NodeInstantiator
+{
+    /** @var string */
+    private $type;
+
+    /** @var ReflectionClass */
+    private $reflectionClass;
+
+    public function __construct(string $type, string $className)
+    {
+        $reflectionClass = new ReflectionClass($className);
+
+        if (! $reflectionClass->isSubclassOf(Node::class)) {
+            throw new InvalidArgumentException(
+                sprintf('%s class is not a subclass of %s', $className, Node::class)
+            );
+        }
+
+        $this->type            = $type;
+        $this->reflectionClass = $reflectionClass;
+    }
+
+    public function getType() : string
+    {
+        return $this->type;
+    }
+
+    /**
+     * @param mixed[] $arguments
+     */
+    public function create(array $arguments) : Node
+    {
+        /** @var Node $node */
+        $node = $this->reflectionClass->newInstanceArgs($arguments);
```

Comment:
i think because we don't know the number of arguments each type of node requires - this way we could pass it as an array

Reply:
I think php allows you to do this other ways now without having to use call_user_func_array?

My thought process for using reflection was that it felt cleaner and I can refactor a little more to control the ReflectionClass instance in a unit test.

Reply:
I changed it to this if that is our preference instead of using reflection.

```
/** @var Node $node */
$node = new $this->className(... $arguments);
```